File: SproInstall/SocketPro/tutorial/CSharp/SampleThree/Client/TOne.cs


using System;
using SocketProAdapter;
using SocketProAdapter.ClientSide;
using USOCKETLib;

public class CTOne : CAsyncServiceHandler
{
    public CTOne()
        : base(TOneConst.sidCTOne)
    {
    }

    public CTOne(CClientSocket cs)
        : base(TOneConst.sidCTOne, cs)
    {
    }

    public CTOne(CClientSocket cs, IAsyncResultsHandler DefaultAsyncResultsHandler)
        : base(TOneConst.sidCTOne, cs, DefaultAsyncResultsHandler)
    {
    }

    public int QueryCount()
    {
        int n;
        bool b = ProcessR1(TOneConst.idQueryCountCTOne, out n);
        return n;
    }

    public int QueryGlobalCount()
    {
        int n;
        bool b = ProcessR1(TOneConst.idQueryGlobalCountCTOne, out n);
        return n;
    }

    public int QueryGlobalFastCount()
    {
        int n;
        bool b = ProcessR1(TOneConst.idQueryGlobalFastCountCTOne, out n);
        return n;
    }

    public void Sleep(int nTime)
    {
        bool b = ProcessR0(TOneConst.idSleepCTOne, nTime);
    }

    public object Echo(object objInput)
    {
        object obj;
        int[] groups = { 1, 2 };
        //send a text message to group 1, SOneSvs Group
        bool b = GetAttachedClientSocket().Push.Broadcast("EchoData called", groups);
        b = ProcessR1(TOneConst.idEchoCTOne, objInput, out obj);
        return obj;
    }

    public void GetAllCounts(out int nCount, out int nGlobalCount, out int nGlobalFastCount)
    {
        int n = 0, global = 0, fast = 0;
        BeginBatching();
        SendRequest(TOneConst.idQueryCountCTOne, delegate(CAsyncResult ar)
        {
            ar.UQueue.Pop(out n);
        });
        SendRequest(TOneConst.idQueryGlobalCountCTOne, delegate(CAsyncResult ar)
        {
            ar.UQueue.Pop(out global);
        });
        SendRequest(TOneConst.idQueryGlobalFastCountCTOne, delegate(CAsyncResult ar)
        {
            ar.UQueue.Pop(out fast);
        });
        CommitBatch(true); //true -- ask server to send three results back in one batch
        WaitAll();
        nCount = n;
        nGlobalCount = global;
        nGlobalFastCount = fast;
    }

    public void GetAllCounts(out int nCount, out int nGlobalCount)
    {
        int n = 0, global = 0;
        BeginBatching();
        SendRequest(TOneConst.idQueryCountCTOne, delegate(CAsyncResult ar)
        {
            ar.UQueue.Pop(out n);
        });
        SendRequest(TOneConst.idQueryGlobalCountCTOne, delegate(CAsyncResult ar)
        {
            ar.UQueue.Pop(out global);
        });
        CommitBatch(true); //true -- ask server to send two results back in one batch
        WaitAll();
        nCount = n;
        nGlobalCount = global;
    }

    public void GetAllCounts(out int nCount)
    {
        nCount = QueryCount();
    }
}

File: SproInstall/SocketPro/samples/JavaScript/tone.js

CMyAsyncHandler = function(){
    var m_idSvsID = 268435476; //tutorial one service id
    var idQueryCountCTOne = 8193;
    var idQueryGlobalCountCTOne = 8194;
    var idQueryGlobalFastCountCTOne = 8195;
    var idSleepCTOne = 8196;
    var idEchoCTOne = 8197;
    
    var me = new SocketProAdapter.ClientSide.CAsyncServiceHandler(); //inheriet from the base CAsyncServiceHandler
    me.GetSvsID = function(){ //must define the method GetSvsID()
        return m_idSvsID; 
    };
    me.OnResultReturned = function(nRequestID, jsQueue){ //must define the callback OnResultReturned(nRequestID, jsQueue)
        switch(nRequestID){
        case idQueryCountCTOne:
            m_QueryCountRtn = jsQueue.loadInt32();
            break;
        case idQueryGlobalCountCTOne:
            m_QueryGlobalCountRtn = jsQueue.loadInt32();
            break;
        case idQueryGlobalFastCountCTOne:
            m_QueryGlobalFastCountRtn = jsQueue.loadInt32();
            break;
        case idSleepCTOne:
            break;
        case idEchoCTOne:
            m_EchoRtn = jsQueue.load();
            break;
        default:
            break;
        }
    };
    
    me.m_QueryCountRtn = 0; //int
	me.QueryCountAsyn = function(){
	    var queue = me.GetAttachedClientSocket().GetMemoryQueue();
	    queue.size = 0;
		return me.SendRequest(idQueryCountCTOne);
	};
	me.m_QueryGlobalCountRtn = 0; //int
	me.QueryGlobalCountAsyn = function(){
	    var queue = me.GetAttachedClientSocket().GetMemoryQueue();
	    queue.size = 0;
		return me.SendRequest(idQueryGlobalCountCTOne);
	};
	me.m_QueryGlobalFastCountRtn = 0; //int
	me.QueryGlobalFastCountAsyn = function(){
	    var queue = me.GetAttachedClientSocket().GetMemoryQueue();
	    queue.size = 0;
		return me.SendRequest(idQueryGlobalFastCountCTOne);
	};
	me.SleepAsyn = function(nTime){
		var queue = me.GetAttachedClientSocket().GetMemoryQueue();
	    queue.size = 0;
	    queue.saveInt32(nTime);
		return me.SendRequest(idSleepCTOne);
	}
    	me.m_EchoRtn = 0; //variant
	me.EchoAsyn = function(objInput) //variant
	{
	    m_EchoRtn = null;
		var queue = me.GetAttachedClientSocket().GetMemoryQueue();
	    queue.size = 0;
	    queue.save(objInput);
		var b = me.SendRequest(idEchoCTOne);
		return b;
	};
	me.QueryCount = function(){
		me.QueryCountAsyn();
		if(me.GetAttachedClientSocket().WaitAll())
		    return m_QueryCountRtn;
		return false;
	};
	me.QueryGlobalCount = function(){
		if(!me.QueryGlobalCountAsyn())
		    return false;
		if(me.GetAttachedClientSocket().WaitAll())
		    return m_QueryGlobalCountRtn;
		return false;
	};
	me.QueryGlobalFastCount = function(){
		if(!me.QueryGlobalFastCountAsyn())
		    return false;
		if(me.GetAttachedClientSocket().WaitAll())
		    return m_QueryGlobalFastCountRtn;
		return false;
	};
	me.Sleep = function(nTime){
		if(!me.SleepAsyn(nTime))
		    return false;
		return me.GetAttachedClientSocket().WaitAll();
	};
	me.Echo = function(objInput){
		if(!me.EchoAsyn(objInput))
		    return false;
		if(me.GetAttachedClientSocket().WaitAll())
		    return m_EchoRtn;
		return false;
	};
	
	me.SendAllRequestsAsyncnronusly = function(sleepTime, echoData){
	    //error handling ignore here
	    me.GetAttachedClientSocket().BeginBatching();
	    me.EchoAsyn(echoData);
	    me.SleepAsyn(sleepTime);
	    me.QueryGlobalCountAsyn();
	    me.QueryCountAsyn();
	    me.QueryGlobalFastCountAsyn();
	    
	    //Send all reuests here, and true -- returning results in batch
	    me.GetAttachedClientSocket().Commit(true);
	    
	    me.GetAttachedClientSocket().WaitAll();
	};
    return me;
};
